### DISCUSSION TITLE
Type 'Props' does not satisfy the constraint 'PageProps' (generateMetadata)

### DISCUSSION THREAD

**Discussion by Dannymx** [vercel/next.js#71997] [Help] [answered]
### Link to the code that reproduces this issue

https://github.com/Dannymx/shockinglemon.com/tree/develop

### To Reproduce

- Clone
- run command 'check-types'
- Will error in any file that has generateMetadata with Props

### Current vs. Expected behavior

Fails at validating types in generateMetadata methods.

### Provide environment information

```bash
❯ npx --no-install next info

Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.0.0: Tue Sep 24 23:37:25 PDT 2024; root:xnu-11215.1.12~1/RELEASE_ARM64_T6030
  Available memory (MB): 36864
  Available CPU cores: 12
Binaries:
  Node: 20.17.0
  npm: 10.8.2
  Yarn: 1.22.22
  pnpm: 9.10.0
Relevant Packages:
  next: 15.0.1 // Latest available version is detected (15.0.1).
  eslint-config-next: 15.0.1
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

TypeScript

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

### Additional context

Only fails in my dynamic pages that generate Metadata info.

```
.next/types/app/band/[slug]/page.ts:34:29 - error TS2344: Type 'Props' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ slug: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

34 checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~

.next/types/app/band/[slug]/page.ts:38:31 - error TS2344: Type 'Props' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ slug: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

38   checkFields<Diff<PageProps, FirstArg<MaybeField<TEntry, 'generateMetadata'>>, 'generateMetadata'>>()
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.next/types/app/blog/[slug]/page.ts:34:29 - error TS2344: Type 'Props' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ slug: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

34 checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~

.next/types/app/blog/[slug]/page.ts:38:31 - error TS2344: Type 'Props' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ slug: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

38   checkFields<Diff<PageProps, FirstArg<MaybeField<TEntry, 'generateMetadata'>>, 'generateMetadata'>>()
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.next/types/app/music/[title]/page.ts:34:29 - error TS2344: Type 'Props' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ title: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

34 checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~

.next/types/app/music/[title]/page.ts:38:31 - error TS2344: Type 'Props' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ title: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

38   checkFields<Diff<PageProps, FirstArg<MaybeField<TEntry, 'generateMetadata'>>, 'generateMetadata'>>()
```

**Comment by samcx** [comment, marked as answer]
This is part of the breaking changes that came with v15.

- https://nextjs.org/docs/app/api-reference/functions/generate-metadata#generatemetadata-function
- https://nextjs.org/blog/next-15#async-request-apis-breaking-change

**Reply by JsemOlik** [reply]
I love you so god damn much. I was trying to figure this out for the past hour and a half 😭

**Reply by vijayluxmimaddheshiya** [reply]
Hi , I am facing same issue but nothing is working for me would it be possible for you to help me out  , thanks!!

**Reply by derekc00** [reply]
Thank u so much :)

**Reply by 1Kunalvats9** [reply]
thanks a lot broooo

**Reply by victorex27** [reply]
This helped me a lot. Thanks

**Reply by yassinl** [reply]
thanks but why did they ship something knowing its gonna break all of our codebases ? are they cruel ? is this a conspiracy to keep developers with work to do ?

**Reply by patrickaigbogun** [reply]
@yassinl i completely relate with and support your frustration.. i feel this should be something that should be introduced within version 14 of nextjs first

**Reply by ilhemmissaoui** [reply]
// src/app/blog-details/[id]/page.tsx

import { Metadata } from "next";
import { notFound } from "next/navigation";
import Wrapper from "@/layouts/Wrapper";
import BlogDetails from "@/components/blog-details";

const blogs = [
  {
    id: "1",
    title: "SEO Service",
    img: "/assets/img/blogs/seo.jpg",
    subTitle: "Get to the limelight with Professional SEO Services",
    desc: "At Sole Digital Media, we rely on data-driven insights to maximize your SEO success...",
    content: `A new digital strategy is putting SEO in the spotlight...`,
  },
  {
    id: "2",
    title: "Branding Design",
    img: "/assets/img/blogs/1.jpg",
    subTitle: "Boost Your Brand With Expert Social Media Marketing Services",
    desc: "One of the biggest benefits of hiring expert services is having full access to detailed analytics...",
    content: `Social media has transformed how businesses engage with customers...`,
  },
  {
    id: "3",
    title: "Web Development",
    img: "/assets/img/blogs/dev.jpg",
    subTitle: "Transform Your Online Presence with Advanced Web Development Services",
    desc: "At SoleDigitalMedia, we are a custom Web Development Company dedicated to the creation...",
    content: `A strong web presence is crucial in the digital age...`,
  },
  {
    id: "4",
    title: "Content Marketing Service",
    img: "/assets/img/blogs/4.JPG",
    subTitle: "Build Strong Customer Relationships with Comprehensive Content Marketing Services",
    desc: "The flow of fresh and high-quality content acts as a pacemaker to the audience...",
    content: `Content marketing focuses on creating valuable and relevant content...`,
  },
];

export async function generateStaticParams() {
  return blogs.map((blog) => ({ id: blog.id }));
}

export async function generateMetadata({
  params,
}: {
  params: { id: string };
}): Promise<Metadata> {
  const blog = blogs.find((b) => b.id === params.id);

  if (!blog) {
    return {
      title: "Not Found - Sole Digital",
      description: "The requested blog was not found.",
    };
  }

  return {
    title: `${blog.title} - Sole Digital`,
    description: blog.subTitle,
  };
}

export default function BlogDetailsPage({
  params,
}: {
  params: { id: string };
}) {
  const blog = blogs.find((b) => b.id === params.id);
  if (!blog) return notFound();

  return (
    <Wrapper>
      <BlogDetails blog={blog} />
    </Wrapper>
  );
}




i got this error please help me to fix it 



.next/types/app/blog-details/[id]/page.ts:34:29
Type error: Type '{ params: { id: string; }; }' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ id: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

  32 |
  33 | // Check the prop type of the entry function
> 34 | checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
     |                             ^
  35 |
  36 | // Check the arguments and return type of the generateMetadata function
  37 | if ('generateMetadata' in entry) {
Next.js build worker exited with code: 1 and signal: null

**Comment by ignatiosdev** [comment]
I fixed it by converting params to a Promise

For example before i did this

```
export default async function RootLayout({
  children,
  params
}: {
  children: React.ReactNode;
  params: {locale: string as "es" | "en"};
}) {
  const { locale } = await params; 
```



Now I do this

```
type Params = Promise<{locale: "en" | "es" }>;

export default async function RootLayout({
  children,
  params
}: {
  children: React.ReactNode;
  params: Params;
}) {
  const { locale } = await params; 
```

**Reply by nathomadri** [reply]
Thank you brother, finally I got the "antidote"

**Reply by QuickPrimeTech** [reply]
Cheers mate

**Comment by mohsinparvi** [comment]
### Here’s how I fix this issse:

```tsx
import Link from "next/link";
import photos, { Photo } from "@/lib/";
import PhotoCard from "@/components/photo-card";
export const generateStaticParams = () => {
  return photos.map(({ id }) => ({
    id: String(id),
  }));
};
export type paramsType = Promise<{ id: string }>;

export default async function PhotoPage(props: { params: paramsType }) {
  const { id } = await props.params;
  const photo: Photo | undefined = photos.find((p) => p.id === id);

  if (!photo) {
    return <div>Photo not found</div>;
  }

  return (
    <section className="py-24">
      <div className="container">
        <div>
          <Link
            href="/photos"
            className="font-semibold italic text-sky-600 underline"
          >
            Back to photos
          </Link>
        </div>
        <div className="mt-10 w-1/3">
          <PhotoCard photo={photo} />
        </div>
      </div>
    </section>
  );
}

**Comment by nxtrs2** [comment]
```
export default async function ArticlePage(props: {
  params: Promise<{ id: string }>;
}) {
  const { id } = await props.params;
  const article = await getArticle(id);
```
  
  so what am I doing wrong? The error won't go away for me.

**Reply by IdrisKulubi** [reply]
npx @next/codemod@canary next-async-request-api .     this will auto update according to the new nextjs 14+

**Comment by 69killian** [comment]
import { format } from 'date-fns';
import prismadb from '@/prisma/prismadb';
import { BillboardClient } from './components/client';
import { BillboardColumn } from './components/columns';

// Définition du type pour params comme une promesse
type Params = Promise<{ storeId: string }>;

// Composant avec params comme Promise
const BillboardsPage = async ({ params }: { params: Params }) => {
    // Attente de la résolution de params
    const { storeId } = await params;

    // Récupération des billboards depuis la base de données
    const billboards = await prismadb.billboard.findMany({
        where: {
            storeId, // Utilisation de storeId après sa résolution
        },
        orderBy: {
            createdAt: "desc", // Tri par date de création décroissante
        },
    });

    // Transformation des données pour correspondre au type BillboardColumn
    const formattedBillboards: BillboardColumn[] = billboards.map((item) => ({
        id: item.id,
        label: item.label,
        createdAt: format(item.createdAt, "MMMM do, yyyy"), // Formatage de la date
    }));

    return (
        <div className="flex-col">
            <div className="flex-1 space-y-4 p-8 pt-6">
                {/* Passer les données formatées au composant BillboardClient */}
                <BillboardClient data={formattedBillboards} />
            </div>
        </div>
    );
};

export default BillboardsPage;
 I also tried, but nothing works for me

**Comment by frankiefab100** [comment]
Here’s how I fix this issue:

Previously it looked like this:
```ts
export default async function Page({
  params,
}: {
  params: { slug: string };
}) {
  const slug = params.slug;
  const response = await getProjectBySlug(slug);
```


Then, I changed it to:

```ts
export default async function Page(props: {
  params: Promise<{ slug: string }>;
}) {
  const params = await props.params;
  const response = await getProjectBySlug(params.slug);
```

**Reply by Tai-DT** [reply]
> Đây là cách tôi khắc phục sự cố này:
> 
> Trước đây nó trông như thế này:
> 
> ```ts
> export default async function Page({
>   params,
> }: {
>   params: { slug: string };
> }) {
>   const slug = params.slug;
>   const response = await getProjectBySlug(slug);
> ```
> 
> Sau đó, tôi đã đổi nó thành:
> 
> ```ts
> export default async function Page(props: {
>   params: Promise<{ slug: string }>;
> }) {
>   const params = await props.params;
>   const response = await getProjectBySlug(params.slug);
> ```

tôi đã thử cách anh và đã thành công cảm ơn anh nhiều

**Reply by ped4enko** [reply]
Thank you, it helped me with the thenudeme.com project

**Comment by rodrigobutta** [comment]
Theres a script that fixes the Params breaking chamnge in the project, the easiest solution:

First commit an leade stage clean, then run `npx @next/codemod@canary next-async-request-api .`

Refer to: https://nextjs.org/blog/next-15#async-request-apis-breaking-change

**Comment by puneethVarma001** [comment]
import ProductListing from "../../components/product-listing";
import { GET_SEARCH_DETAILS } from "../../utils/constants";
interface SearchParams {
  page?: string;
  count?: string;
}
export default async function PLP({ params, searchParams }: Readonly<{ children: React.ReactNode; params: Promise<{ id: number }>, searchParams: SearchParams }>) {
  const { id } = await params;

  const currentPage = searchParams.page ? parseInt(searchParams.page) - 1 : 0;
  const count = searchParams.count ? parseInt(searchParams.count) : 9;

  const fetchCategoryDetails = async (id: number) => {
    try {
      const response = await fetch(
        `https://platform.cutiskart.com/catalog/brand/${id}`,
        {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            deviceType: "WEB",
            Authorization: "Basic Z2Vla3Nfb2ZfYWRoYXJ2YXNvZnQ=",
          },
          //   cache: "no-store",
        }
      );
      console.log("response", response);

      if (!response.ok) return { data: { brandList: [] } };
      return await response.json();
    } catch (error) {
      console.error("Error fetching banners:", error);
      return { data: { brandList: [] } };
    }
  };

  const getPlp = async (id: number, currentPage: number, count: number) => {
    try {
      const response = await fetch(
        `https://platform.cutiskart.com/${GET_SEARCH_DETAILS}?type=2&brandId=${id}&page=${currentPage}&count=${count}&filters=true`,
        {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            deviceType: "WEB",
            Authorization: "Basic Z2Vla3Nfb2ZfYWRoYXJ2YXNvZnQ=",
          },
        }
      );

      if (!response.ok) return { data: { products: [] } };
      return await response.json();
    } catch (error) {
      console.error("Error fetching banners:", error);
      return { data: { products: [] } };
    }
  };

  const breadcrumbData = await fetchCategoryDetails(id);
  const productData = await getPlp(id, currentPage, count);

  return (

    <div>
      <ProductListing
        breadcrumbData={breadcrumbData?.data}
        productData={productData}
        currentPage={currentPage}
        count={count}
        isCategory={false}
      />
    </div>
  )
}

while building it is showing an error given below
app/brand/[id]/page.tsx
Type error: Type 'PageProps' does not satisfy the constraint 'import("/home/tharun/Desktop/cutiskart-new/.next/types/app/brand/[id]/page").PageProps'.
  Types of property 'params' are incompatible.
    Type '{ id: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

Next.js build worker exited with code: 1 and signal: null

**Comment by Vic-Orlands** [comment]
Shit sucks fr!! but this fixed it for me:

type paramsType = Promise<{ slug: string }>;

export async function generateMetadata({
  params,
}: { params: paramsType; ): Promise<Metadata> {
  const { slug } = await params;
  const post = await getBlogPost(slug);
  
  ....other codes here

**Comment by mundane799699** [comment]
I'm using next.js 15.2.0. I solved the problem like this:
``` js
interface ChatIdPageProps {
  // use promise
  params: Promise<{ chatId: string }>;
}

const ChatIdPage = async ({ params }: ChatIdPageProps) => {
  const { userId } = await auth();
  if (!userId) {
    return <RedirectToSignIn />;
  }

  // get the chatId from promise
  const { chatId } = await params;
  const companion = await prismadb.companion.findUnique({
    where: {
      id: chatId,
    },
......
```

**Reply by SouleymaneSy7** [reply]
Thanks you so much. This fixed the issue for me

**Reply by winayaid** [reply]
Thank you this is work for me

**Comment by freddy24-7** [comment]
This worked for me:
const locales = ['en', 'nl', 'tr', 'ar'];

// Define props so that params is always a Promise.
type Props = {
  children: ReactNode;
  params: Promise<{ locale: string }>;
};

export default async function LocaleLayout({ children, params }: Props) {
  // Await params, which is now guaranteed to be a Promise.
  const { locale } = await params;

  if (!locales.includes(locale)) {
    notFound();
  }

**Comment by yaolifeng0629** [comment]
Regarding this issue, the official has provided me with a solution:

Official document: https://nextjs.org/docs/messages/sync-dynamic-apis

Post: https://x.com/Immerse_code/status/1910898542815961248

![image](https://github.com/user-attachments/assets/6cd7cddd-fbff-4a4e-a695-389173f62915)
![image](https://github.com/user-attachments/assets/a22fcda2-832a-4d9a-a40b-176cf390a61f)

**Comment by ttttttomas** [comment]
import { Metadata } from 'next';

type Props = {
  params: {
    periodo: string;
    id: string;
  };
};


export async function generateStaticParams({ params }: { params: Props }): Promise<Metadata> {
  return {title: `${params.params.periodo} - ${params.params.id}`}
}

export default async function ProductPage(props: { params: Props }) {
    const { params } =  props.params;
  const response = await fetch(`{apiURL}/products/${params.id}`); //apiURL IS PRIVATE
  const currentProduct = await response.json();

why do i get this error? : 

![image](https://github.com/user-attachments/assets/fa17e042-496f-4739-871d-442fcff0e100)

**Reply by patrickaigbogun** [reply]
@ttttttomas you have to set the correct interface, then destruct it within your code
Here's an image example


![Screenshot_20250521-100238.jpg](https://github.com/user-attachments/assets/0edb4df1-c998-4ccd-860e-de0187a83ba1)

**Reply by subarnapkrl** [reply]
`
interface GalleryDetailsPageProps {
params: {
id: string;
};
searchParams: {
success?: string;
canceled?: string;
error?: string;
};
}
export default function Page({
params,
searchParams,
}: GalleryDetailsPageProps) {
return (
<Suspense
fallback={



}
>


);
}

`

HELLLLLLLPPPPPPPPP: when npm run build , I tried all the above solutions, but the issue is still not yet fixed
.next/types/app/gallery/[id]/page.ts:34:29
Type error: Type 'GalleryDetailsPageProps' does not satisfy the constraint 'PageProps'.
Types of property 'params' are incompatible.
Type '{ id: string; }' is missing the following properties from type 'Promise': then, catch, finally, [Symbol.toStringTag]

32 |
33 | // Check the prop type of the entry function

    34 | checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
    | ^
    35 |
    36 | // Check the arguments and return type of the generateMetadata function
    37 | if ('generateMetadata' in entry) {
    Next.js build worker exited with code: 1 and signal: null

**Comment by Lucifergene** [comment]
**Note:** If your file uses React Hooks, you can't directly use async/await in that component, as React Hooks require a "use client" directive, whereas async functions are treated as server-side.

If you run into this conflict, a good approach is to nest a client component inside a server component. The server component can handle the data fetching using async/await, and then pass the retrieved values as props to the client component, which can safely use React Hooks.

**Reply by subarnapkrl** [reply]
`
interface GalleryDetailsPageProps {
params: {
id: string;
};
searchParams: {
success?: string;
canceled?: string;
error?: string;
};
}
export default function Page({
params,
searchParams,
}: GalleryDetailsPageProps) {
return (
<Suspense
fallback={



}
>


);
}

`

HELLLLLLLPPPPPPPPP: when npm run build , I tried all the above solutions, but the issue is still not yet fixed
.next/types/app/gallery/[id]/page.ts:34:29
Type error: Type 'GalleryDetailsPageProps' does not satisfy the constraint 'PageProps'.
Types of property 'params' are incompatible.
Type '{ id: string; }' is missing the following properties from type 'Promise': then, catch, finally, [Symbol.toStringTag]

32 |
33 | // Check the prop type of the entry function

    34 | checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
    | ^
    35 |
    36 | // Check the arguments and return type of the generateMetadata function
    37 | if ('generateMetadata' in entry) {
    Next.js build worker exited with code: 1 and signal: null

**Comment by subarnapkrl** [comment]
`
interface GalleryDetailsPageProps {
  params: {
    id: string;
  };
  searchParams: {
    success?: string;
    canceled?: string;
    error?: string;
  };
}
export default function Page({
  params,
  searchParams,
}: GalleryDetailsPageProps) {
  return (
    <Suspense
      fallback={
        <div className="flex items-center justify-center min-h-[65vh]">
          <Loader2 className="h-8 w-8 animate-spin text-black" />
        </div>
      }
    >
      <GalleryContent params={params} searchParams={searchParams} />
    </Suspense>
  );
}

`

HELLLLLLLPPPPPPPPP: when npm run build , I tried all the above solutions, but the issue is still not yet fixed
.next/types/app/gallery/[id]/page.ts:34:29
Type error: Type 'GalleryDetailsPageProps' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ id: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]

  32 |
  33 | // Check the prop type of the entry function
> 34 | checkFields<Diff<PageProps, FirstArg<TEntry['default']>, 'default'>>()
     |                             ^
  35 |
  36 | // Check the arguments and return type of the generateMetadata function
  37 | if ('generateMetadata' in entry) {
Next.js build worker exited with code: 1 and signal: null

**Comment by patrickaigbogun** [comment]
Aiit
